connectors/byoc.py

```diff
@@ -498,77 +430,78 @@ def status(self, value):
+    def sync_now(self):
+        return self.get("sync_now", default=False)
+
+    @property
+    def scheduling(self):
+        return self.get("scheduling", default={})
+
+    @property
+    def configuration(self):
+        return DataSourceConfiguration(self.get("configuration"))
 
     @configuration.setter
     def configuration(self, value):
-        self.doc_source["configuration"] = value
-        status = (
-            Status.CONFIGURED
-            if all(
-                isinstance(value, dict) and value.get("value") is not None
-                for value in value.values()
-            )
-            else Status.NEEDS_CONFIGURATION
-        )
-        self.doc_source["status"] = status.value
-        self._update_config(self.doc_source)
-
-    async def suspend(self):
-        if self._sync_task is not None:
-            task = self._sync_task
-            task.cancel()
-            await task
-        await self.close()
-
-    async def close(self):
-        self._closed = True
-        if self._heartbeat_started:
-            self._hb.cancel()
-            self._heartbeat_started = False
-        if self.data_provider is not None:
-            await self.data_provider.close()
-            self.data_provider = None
-
-    async def sync_doc(self, force=True):
-        if not self._dirty and not force:
-            return
-        self.doc_source["last_seen"] = iso_utc()
-        await self.index.save(self)
-        self._dirty = False
+        self._source["configuration"] = value
 
-    def start_heartbeat(self, delay):
-        if self._heartbeat_started:
-            return
-        self._heartbeat_started = True
+    @property
+    def index_name(self):
+        return self.get("index_name")
+
+    @property
+    def language(self):
+        return self.get("language")
 
-        async def _heartbeat():
-            while not self._closed:
-                logger.info(f"*** Connector {self.id} HEARTBEAT")
-                if not self._syncing:
-                    self.doc_source["last_seen"] = iso_utc()
-                    await self.sync_doc()
-                await asyncio.sleep(delay)
+    @property
+    def filtering(self):
+        return Filtering(self.get("filtering"))
 
-        self._hb = asyncio.create_task(_heartbeat())
+    @filtering.setter
+    def filtering(self, value):
+        self._source["filtering"] = value
+
+    @property
+    def pipeline(self):
+        return Pipeline(self.get("pipeline"))
+
+    @property
+    def features(self):
+        return Features(self.get("features"))
+
+    @property
+    def last_sync_status(self):
+        return JobStatus(self.get("last_sync_status"))
+
+    async def heartbeat(self, interval):
+        if (
+            self.last_seen is None
+            or (datetime.now(timezone.utc) - self.last_seen).total_seconds() > interval
+        ):
+            logger.info(f"*** Connector {self.id} HEARTBEAT")
+            await self.index.heartbeat(doc_id=self.id)
```

Comment:
I think else with `logger.debug` is good here.

But even better to move the `if` statement to the caller of the heartbeat. IMO this logic should reside outside of the class.

Reply:
> I think else with `logger.debug` is good here.

This was copied from here:
https://github.com/elastic/connectors-python/blob/e75e7b30f5124596151451d296fca7022bb728af/connectors/byoc.py#L565

I keep this because some tests asset presence of this log.


> But even better to move the `if` statement to the caller of the heartbeat. IMO this logic should reside outside of the class.

I can do that.

Reply:
Think twice, I think it also makes sense to let `Connector` handle when to send heartbeat. If we move the `if` statement to the caller(i.e. `SyncService`), it will be `SyncService` which controls when to send heartbeat. And `Connector.heartbeat()` method will simply update `last_seen`.

There is a very similar case in method `prepare` of class `Connector`. The `prepare` method will update `service_type` and `configuration` for `Build a connector` connector for the first time. The `prepare` method controls when to update, and when to return early. The caller (`SyncService`) simply calls `connector.prepare(config)`.

For all the native connectors, and all the subsequent calls (ideally) of `Build a connector` connectors, it will return directly from here:

https://github.com/elastic/connectors-python/blob/544cfea7eeb6bd874592155afd73405a6d917b8d/connectors/byoc.py#L626-L627